// ipysheet/lib/test/test_sheet.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ipysheet = require("../sheet");
const dummy_manager_1 = require("./dummy-manager");
const chai_1 = require("chai");
const utils_1 = require("./utils");
var data_cloner = function () {
    var data = this.sheet.data;
    return JSON.parse(JSON.stringify(data));
};
describe('sheet', function () {
    beforeEach(function () {
        return __awaiter(this, void 0, void 0, function* () {
            this.manager = new dummy_manager_1.DummyManager({ ipysheet: ipysheet });
            const modelId = 'u-u-i-d';
            this.sheet = yield this.manager.new_widget({
                model_module: 'ipysheet',
                model_name: 'SheetModel',
                model_module_version: '*',
                view_module: 'jupyter-widgets',
                view_name: 'DOMWidgetView',
                view_module_version: '*',
                //model_module: 'test-widgets',
                //model_name: 'TestWidget',
                model_id: modelId,
            }, { rows: 2, columns: 4 });
            this.sheet.state_change = Promise.resolve(); // bug in ipywidgets?
            this.sheet.views = {};
        });
    });
    it('sanity', function () {
        chai_1.expect(this.sheet.get('rows')).to.equal(2);
        chai_1.expect(this.sheet.get('columns')).to.equal(4);
    });
    it('data init', function () {
        //console.log(this.sheet.data)
        chai_1.expect(this.sheet.data).to.have.lengthOf(2);
        chai_1.expect(this.sheet.data[0]).to.have.lengthOf(4);
    });
    it('data row grow', function () {
        // test grow
        this.sheet.data[1][2].value = 123;
        this.sheet.set('rows', 3);
        chai_1.expect(this.sheet.data).to.have.lengthOf(3);
        chai_1.expect(this.sheet.data[0]).to.have.lengthOf(4);
        chai_1.expect(this.sheet.data[2]).to.have.lengthOf(4);
        chai_1.expect(this.sheet.data[1][2].value, 'data should be preserved when changing size').to.equal(123);
        this.sheet.set('rows', 5);
        chai_1.expect(this.sheet.data).to.have.lengthOf(5);
        chai_1.expect(this.sheet.data[0]).to.have.lengthOf(4);
        chai_1.expect(this.sheet.data[2]).to.have.lengthOf(4);
        chai_1.expect(this.sheet.data[1][2].value, 'data should be preserved when changing size').to.equal(123);
    });
    it('data column shrink', function () {
        this.sheet.data[1][2].value = 123;
        this.sheet.data[1][3].value = 1234;
        this.sheet.set('columns', 3);
        chai_1.expect(this.sheet.data).to.have.lengthOf(2);
        chai_1.expect(this.sheet.data[0]).to.have.lengthOf(3);
        chai_1.expect(this.sheet.data[1]).to.have.lengthOf(3);
        chai_1.expect(this.sheet.data[1][2].value, 'data should be preserved when changing size').to.equal(123);
    });
    it('data column grow', function () {
        this.sheet.set('columns', 5);
        chai_1.expect(this.sheet.data).to.have.lengthOf(2);
        chai_1.expect(this.sheet.data[0]).to.have.lengthOf(5);
        chai_1.expect(this.sheet.data[1]).to.have.lengthOf(5);
    });
    it('model reflecting view', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var view = yield utils_1.make_view.call(this);
            yield view._table_constructed;
            view.set_cell(1, 2, 123);
            yield utils_1.wait_validate(view);
            chai_1.expect(this.sheet.data[1][2].value, 'cell changes should be reflected in model').to.equal(123);
        });
    });
    it('view reflecting model', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var view = yield utils_1.make_view.call(this);
            var data = this.sheet.data;
            yield view._table_constructed;
            data[1][2].value = 123;
            this.sheet.trigger('data_change');
            yield view._last_data_set;
            chai_1.expect(view.get_cell(1, 2), 'model change should be reflected in view').to.equal(123);
        });
    });
    it('view reflecting different view', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var view1 = yield utils_1.make_view.call(this);
            var view2 = yield utils_1.make_view.call(this);
            view1.set_cell(1, 2, 123);
            var bla = yield utils_1.wait_validate(view1);
            chai_1.expect(view1.get_cell(1, 2), 'model change should be reflected in view').to.equal(123);
            chai_1.expect(view1.get_cell(1, 2), 'cell changes in one view should be reflected in a related view').to.equal(view2.get_cell(1, 2));
        });
    });
    // we don't validate at the moment
    it.skip('invalid sheet should not propagate to model', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var view = yield utils_1.make_view.call(this);
            var data_clone = data_cloner.call(this);
            data_clone[1][2].value = 123;
            data_clone[1][2].options = { type: 'numeric' };
            chai_1.expect(data_clone[1][2].value, 'cloned data check').to.equal(123);
            this.sheet.set('data', data_clone);
            view.set_cell(1, 2, 'wrong');
            yield utils_1.wait_validate(view);
            chai_1.expect(view.get_cell(1, 2), 'sheet will reflect invalid data').to.equal('wrong');
            chai_1.expect(this.sheet.data[1][2].value, 'model should not have invalid data').to.not.equal('wrong');
        });
    });
    var make_cell = function (options, skip_add) {
        return __awaiter(this, void 0, void 0, function* () {
            const modelId = 'u-u-i-d-cell';
            var cell = yield this.manager.new_widget({
                model_module: 'ipysheet',
                model_name: 'CellRangeModel',
                view_module: 'jupyter-widgets',
                view_name: 'DOMWidgetView',
                view_module_version: '*',
                model_module_version: '0.1.0',
                model_id: modelId,
            }, Object.assign({ row_start: 1, column_start: 2, row_end: 1, column_end: 2, value: 888 }, options));
            var cells = this.sheet.get('cells');
            if (!skip_add)
                this.sheet.set('cells', [...cells, cell]);
            return cell;
        });
    };
    var make_range = function (options, skip_add) {
        return __awaiter(this, void 0, void 0, function* () {
            const modelId = 'u-u-i-d-cell';
            var cell = yield this.manager.new_widget({
                model_module: 'ipysheet',
                model_name: 'CellRangeModel',
                view_module: 'jupyter-widgets',
                view_name: 'DOMWidgetView',
                view_module_version: '*',
                model_module_version: '0.1.0',
                model_id: modelId,
            }, Object.assign({ squeeze_row: false, squeeze_column: false }, options));
            var cells = this.sheet.get('cells');
            if (!skip_add)
                this.sheet.set('cells', [...cells, cell]);
            return cell;
        });
    };
    it('cell changes should be reflected in datamodel', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: 777 }]);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'for initial value').to.equal(777);
            cell.set('value', 999);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'when cell.value is change').to.equal(999);
        });
    });
    it('numeric cell with value zero should indeed have value zero', function () {
        return __awaiter(this, void 0, void 0, function* () {
            yield make_cell.apply(this, [{ value: 0.00, type: 'numeric' }]);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'for initial value').to.equal(0);
        });
    });
    it('none cell with should be set', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: 0.00, type: 'numeric' }]);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'for initial value').to.equal(0);
            cell.set('value', null);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'for new value').to.equal(null);
        });
    });
    it('multiple cells added', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell1 = yield make_cell.apply(this, [{ value: 777 }, true]);
            var cell2 = yield make_cell.apply(this, [{ row_start: 0, row_end: 0, value: 555 }, true]);
            var cells = this.sheet.get('cells');
            this.sheet.set('cells', [...cells, cell1, cell2]);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'for initial value').to.equal(777);
            chai_1.expect(data[0][2].value, 'for initial value').to.equal(555);
            cell1.set('value', 999);
            cell2.set('value', 444);
            var data = this.sheet.data;
            chai_1.expect(data[1][2].value, 'when cell.value is change').to.equal(999);
            chai_1.expect(data[0][2].value, 'when cell.value is change').to.equal(444);
        });
    });
    it('model changes should be reflected in cell', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: [[777]] }]);
            var data = data_cloner.call(this);
            data[1][2].value = 999;
            this.sheet.data = data;
            this.sheet.trigger('data_change');
            chai_1.expect(cell.get('value'), 'when the data in the sheet changes').to.equal(999);
        });
    });
    it('range should be reflected in all cells', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var range = yield make_range.apply(this, [{ row_start: 0, row_end: 1, column_start: 1, column_end: 2, value: [[0, 1], [2, 3]] }]);
            var data = data_cloner.call(this);
            chai_1.expect(data[0][1].value, 'and in the underlying data grid').to.equal(0);
            chai_1.expect(data[0][2].value, 'and in the underlying data grid').to.equal(1);
            chai_1.expect(data[1][1].value, 'and in the underlying data grid').to.equal(2);
            chai_1.expect(data[1][2].value, 'and in the underlying data grid').to.equal(3);
            var cell = yield make_cell.apply(this, [{ row_start: 0, row_end: 0, value: 777 }]);
            data = data_cloner.call(this);
            chai_1.expect(data[0][2].value, 'sanity check').to.equal(777);
            chai_1.expect(range.get('value')[0][1], 'and in the underlying data grid, synced back').to.equal(777);
            var cell2 = yield make_cell.apply(this, [{ row_start: 1, row_end: 1, value: null }]);
            data = data_cloner.call(this);
            chai_1.expect(data[1][2].value, 'should not be using a null value for an overlapping value').to.equal(3);
            chai_1.expect(range.get('value')[1][1], 'and the original data should not be modified').to.equal(3);
            chai_1.expect(cell2.get('value'), 'but the cell value should be updated').to.equal(3);
            //data[1][2].value = 999;
            //this.sheet.set('data', data)
            //expect(range.get('value'), 'when the data in the sheet changes').to.equal(999);
        });
    });
    it('range tranposed should be reflected in all cells', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var range = yield make_range.apply(this, [{ row_start: 0, row_end: 1, column_start: 1, column_end: 2, transpose: true, value: [[0, 1], [2, 3]] }]);
            var data = data_cloner.call(this);
            chai_1.expect(data[0][1].value, 'and in the underlying data grid').to.equal(0);
            chai_1.expect(data[0][2].value, 'and in the underlying data grid').to.equal(2);
            chai_1.expect(data[1][1].value, 'and in the underlying data grid').to.equal(1);
            chai_1.expect(data[1][2].value, 'and in the underlying data grid').to.equal(3);
            var cell = yield make_cell.apply(this, [{ row_start: 0, row_end: 0, value: 777 }]);
            data = data_cloner.call(this);
            chai_1.expect(data[0][2].value, 'sanity check').to.equal(777);
            chai_1.expect(range.get('value')[1][0], 'and in the underlying data grid, synced back').to.equal(777);
            //data[1][2].value = 999;
            //this.sheet.set('data', data)
            //expect(range.get('value'), 'when the data in the sheet changes').to.equal(999);
        });
    });
    it('should use the last style but not overwrite', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var range = yield make_range.apply(this, [{ row_start: 0, row_end: 1, column_start: 1, column_end: 2, transpose: true, value: [[0, 1], [2, 3]],
                    style: { color: 'red', backgrouncColor: 'orange' } }]);
            var data = data_cloner.call(this);
            chai_1.expect(data[0][1].options.style.color, 'and in the underlying data grid').to.equal('red');
            chai_1.expect(data[0][2].options.style.color, 'and in the underlying data grid').to.equal('red');
            chai_1.expect(data[1][1].options.style.color, 'and in the underlying data grid').to.equal('red');
            chai_1.expect(data[1][2].options.style.color, 'and in the underlying data grid').to.equal('red');
            chai_1.expect(data[0][1].options.style.backgrouncColor, 'and in the underlying data grid').to.equal('orange');
            chai_1.expect(data[0][2].options.style.backgrouncColor, 'and in the underlying data grid').to.equal('orange');
            chai_1.expect(data[1][1].options.style.backgrouncColor, 'and in the underlying data grid').to.equal('orange');
            chai_1.expect(data[1][2].options.style.backgrouncColor, 'and in the underlying data grid').to.equal('orange');
            var cell = yield make_cell.apply(this, [{ row_start: 0, row_end: 0, value: 777, style: { color: 'blue' } }]);
            data = data_cloner.call(this);
            chai_1.expect(data[0][2].options.style.color, 'effective color should be blue').to.equal('blue');
            chai_1.expect(data[0][2].options.style.backgrouncColor, 'effective backgrouncColor should be blue').to.equal('orange');
            chai_1.expect(range.get('style').color, 'but the original should not be changed').to.equal('red');
        });
    });
    it('should search at table creation', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: [['Hello']] }]);
            this.sheet.set('search_token', 'Hell');
            var view = yield utils_1.make_view.call(this);
            yield view._table_constructed;
            chai_1.expect(view.el.querySelector('td[class*="htSearchResult"]')).to.not.equal(null);
        });
    });
    it('should search', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: [['Hello']] }]);
            var view = yield utils_1.make_view.call(this);
            yield view._table_constructed;
            chai_1.expect(view.el.querySelector('td[class*="htSearchResult"]')).to.equal(null);
            this.sheet.set('search_token', 'Hell');
            chai_1.expect(view.el.querySelector('td[class*="htSearchResult"]')).to.not.equal(null);
        });
    });
    it('should search after change', function () {
        return __awaiter(this, void 0, void 0, function* () {
            var cell = yield make_cell.apply(this, [{ value: [['Yop']] }]);
            var view = yield utils_1.make_view.call(this);
            yield view._table_constructed;
            this.sheet.set('search_token', 'Hell');
            chai_1.expect(view.el.querySelector('td[class*="htSearchResult"]')).to.equal(null);
            cell.set('value', [['Hello']]);
            yield view._last_data_set;
            chai_1.expect(view.el.querySelector('td[class*="htSearchResult"]')).to.not.equal(null);
        });
    });
});
